warehouse/packaging/models.py

```diff
@@ -330,7 +331,11 @@ def __acl__(self):
         # The project has zero or more OIDC publishers registered to it,
         # each of which serves as an identity with the ability to upload releases.
         for publisher in self.oidc_publishers:
-            acls.append((Allow, f"oidc:{publisher.id}", [Permissions.ProjectsUpload]))
+            if self.lifecycle_status != LifecycleStatus.Archived:
+                # Only allow uploads in non-archived projects
+                acls.append(
+                    (Allow, f"oidc:{publisher.id}", [Permissions.ProjectsUpload])
+                )
```

Comment:
optimize: Adding the `if` will now happen on each iteration of the loop per `publisher`, but the project's lifecycle happens once, probably can shortcut the loop instead by placing the `if` outside.
Most projects only have a single publisher, but just in case, save the iteration if unnecessary.